torchaudio_contrib/functional.py

```diff
@@ -1,106 +1,133 @@
 import torch
-import numpy as np
+import math
+import torch.nn.functional as F
 
 
-def _get_time_values(sig_length, sr, hop):
+def stft_defaults(n_fft, hop_length, len_win, window):
     """
-    Get the time axis values given the signal length, sample
-    rate and hop size.
+    Should function outisde torchaudio_contrib.stft since
+    torchaudio_contrib.STFT will use it outisde of forward() (?).
     """
-    return torch.linspace(0, sig_length/sr, sig_length//hop+1)
+    hop_length = n_fft // 4 if hop_length is None else hop_length
 
+    if window is None:
+        length = n_fft if len_win is None else len_win
+        window = torch.hann_window(length)
```

Comment:
I think creating the window in here may not be in line with the spirit of `torch.nn.functional`. For 2d convolution, for example, the class (`nn.Conv2d`) instantiates the weights, and the function (`F.conv2d`) only performs the operation. So I think all operators should be passed as parameters; not created inside a function. I think the idea of the functional interface is just to make it easier to reuse operations in different classes, but it should not encourage inefficient forward passes.

Reply:
Right... I guess from our conversations with @keunwoochoi and @faroit it seemed that we wanted the functionals to also be able to handle the same operations as the layers. Especially if we're doing composition, `layers.Spectrogram` composes `layers.STFT` and `layers.ComplexNorm` and doesn't make `functional.spectrogram` perform the operation (I guess it makes `functional._stft` do it). We could have `functional.stft` take in the window as an argument and not create it (so in essence what `_stft` does) like you've mentioned below and then the case does seem analogous to the `nn.Conv2d`? 
I think the question is if there's a use case for someone wanting to compute the spectrogram (or whatever transform) outside of the `nn.Module` and do we want to support that.

Even more apparent with `functional.melspectrogram`, where "it's doing none of the work" of `layers.Melspectrogram`. If we follow the above procedure with `melspectrogram` it would take as input the spectrogram tensor and also the mel_filterbank, and it would only be multiplying them, so becomes kinda pointless... ?

So I think a good discussion is: Do we want the functionals to carry out the same operations as the layers (kinda how it's described in the README, where e.g. `Spectrogram` and `spectrogram` have the same arguments). Or do we want them to perform the operations of the layer (in this case we would just need the `stft` that takes as input the window buffer, the `complex_norm` and `create_mel_filter`, and not `spectrogram` or `melspectrogram`).

Thougths?

Reply:
> If we follow the above procedure with `melspectrogram` it would take as input the spectrogram tensor and also the mel_filterbank, and it would only be multiplying them, so becomes kinda pointless... ?

Well, it would do the necessary transpose to move the bands to the end before the multiplication and move them back afterwards. Similarly, `stft` would be a thin wrapper around `torch.stft` just to add support for any dimensionality. If `torch.stft` would provide this by itself, then our `STFT` would just call `torch.stft` and we would not provide a functional version at all.

> So I think a good discussion is: Do we want the functionals to carry out the same operations as the layers [...]. Or do we want them to perform the operations of the layer [...].

I think we should follow `torch.nn`, so that would be the second option. But I don't know what you discussed before -- @keunwoochoi and @faroit (when you're back), what did you think the functional interface would be used for? I'll open a separate issue for this.

> in this case we would just need the `stft` that takes as input the window buffer, the `complex_norm` and `create_mel_filter`, and not `spectrogram`

True, maybe we wouldn't duplicate `spectrogram` as doing the same composition that `Spectrogram` does. We should probably only have the composition once. But maybe `Spectrogram` should call `spectrogram` and not use `nn.Sequential`? Ah no, then `Spectrogram` would have to recreate the logic of holding the `window`, or derive from `STFT` (which is what I thought we should avoid, in case we want to do, say, a wavelet transform instead of the stft and still use the same complex norm).

>  or `melspectrogram`).

Instead of `melspectrogram`, we might just have a more generic `apply_filterbank` that would handle the transpose + matmul + transpose. (But `apply_filterbank` is not a good name, this doesn't make clear it's in the spectral domain.)

Reply:
> Ah no, then Spectrogram would have to recreate the logic of holding the window

:+1: 

> I think we should follow torch.nn, so that would be the second option

:+1: If everyone else agrees I'll make the changes! Let's continue discussing in [35](https://github.com/keunwoochoi/torchaudio-contrib/issues/35).

> Instead of melspectrogram, we might just have a more generic apply_filterbank that would handle the transpose + matmul + transpose.

Right! Any ideas for the name? :)